src/server/editorServices.ts

```diff
@@ -754,6 +756,103 @@ namespace ts.server {
+
+            // Todo: add support for local cache path
+            const cachePath = this.host.globalTypingCachePath;
+            const tsdJsonPath = toPath("tsd.json", cachePath, getCanonicalFileName);
+            const tsdJson = this.host.getTsdJson(tsdJsonPath);
+            const { cachedTypingPaths, newTypingNames } = JsTyping.discoverTypings(
+                sys,
+                /* fileNames */ project.getFileNames(),
+                /* projectRootPath */ projectRootPath,
+                /* safeListPath */ undefined,
+                /* packageNameToTypingLocation */ tsdJson ? tsdJson.packageNameToTypingPath : undefined,
+                typingOptions,
+                compilerOptions
+            );
+
+            this.log("Cached typings include: " + cachedTypingPaths);
+            this.log("New typings to download: " + newTypingNames);
+
+            // Bail out when no actions are needed
+            if (cachedTypingPaths.length === 0 && newTypingNames.length === 0) {
+                return;
+            }
+
+            if (!sys.directoryExists(cachePath)) {
+                sys.createDirectory(cachePath);
+            }
+
+            // Inject the cached ones first
+            cachedTypingPaths.forEach(p => addTypingToProject(p, project));
+
+            if (newTypingNames && newTypingNames.length > 0) {
+                const typingsFolderPath = toPath("typings", this.host.globalTypingCachePath, getCanonicalFileName);
+                const query = new tsd.Query();
+                const api = tsd.getAPI(tsdJsonPath);
+                const options = new tsd.Options();
+                options.resolveDependencies = true;
+                options.overwriteFiles = true;
+                options.saveToConfig = true;
+
+                for (const newTypingName of newTypingNames) {
+                    query.addNamePattern(newTypingName);
+                }
+
+                let promise: any;
+                if (!sys.fileExists(tsdJsonPath)) {
+                    promise = api.initConfig(/*overwrite*/ true).then((paths: string[]) => api.readConfig());
+                }
+                else {
+                    promise = api.readConfig();
+                }
+
+                promise
+                    .then(() => api.select(query, options))
+                    .then((selection: any) => api.install(selection, options))
+                    .then((installResult: any) => {
+                        const keys = getKeys(installResult.written.dict);
+                        for (const key of keys) {
+                            const installedTypingPath = toPath(key, typingsFolderPath, getCanonicalFileName);
+                            // If the tsd.json was just created (after calling the init api), then
+                            // the it is not cached yet, the `tsdJson` variable will be undefined

```

Comment:
By "not cached" I meant that the local variable `tsd` would be `undefined`, therefore the checking in the if block is needed.